### PR TITLE
Adds OpenAI functions powered structured output chain

### PR DESCRIPTION
**Pull request by jacoblee93**
Adds chain that leverages OpenAI functions to structure output rather than stuffing format instructions into the prompt. Should help with reliability for more complex schemas.

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
looks pretty reasonable (high level) to me

File: langchain/src/chains/openai_functions/structured_output.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  input: FunctionCallStructuredOutputChainInput<T>
+) {
+  const {
+    outputSchema,
+    llm = new ChatOpenAI({ modelName: "gpt-3.5-turbo-0613", temperature: 0 }),
+    outputKey = "output",
+    llmKwargs = {},
+    ...rest
+  } = input;
+  const functionName = "__lc_output__";
```
Comment (baskaryan) [inline langchain/src/chains/openai_functions/structured_output.ts:63]: would it be better to give a more descriptive name to the model? like output_formatter

File: langchain/src/chains/openai_functions/structured_output.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        `Failed to parse. Text: "${initialParseResult}". Error: ${e}`,
+        initialParseResult
+      );
+    }
+  }
+}
+
+/**
+ * Create a chain for querying an API from a OpenAPI spec.
+ * @param spec OpenAPISpec or url/file/text string corresponding to one.
```
Comment (baskaryan) [inline langchain/src/chains/openai_functions/structured_output.ts:49]: does this docstring apply?

Comment (jacoblee93) [reply langchain/src/chains/openai_functions/structured_output.ts:49]: It does not, will fix.

File: langchain/src/chains/openai_functions/structured_output.ts (hunk 1 above)
Comment (jacoblee93) [reply langchain/src/chains/openai_functions/structured_output.ts:63]: Will do!

**Pull request merged**